crates/comms/src/livekit_web.rs

```diff
@@ -0,0 +1,279 @@
+    sender: Sender<PlayerUpdate>,
+) -> Result<(), anyhow::Error> {
+    let room = connect_room(address, token)
+        .await
+        .map_err(|e| anyhow::anyhow!("Failed to connect room: {:?}", e))?;
+
+    let sender_clone = sender.clone();
+
+    // Set up event handler
+    let event_handler = Closure::wrap(Box::new(move |event: JsValue| {
+        let sender = sender_clone.clone();
+
+        spawn_local(async move {
+            handle_room_event(event, transport_id, sender).await;
+        });
+    }) as Box<dyn FnMut(JsValue)>);
+
+    set_room_event_handler(&room, &event_handler)
+        .map_err(|e| anyhow::anyhow!("Failed to set event handler: {:?}", e))?;
+
+    // Keep the closure alive
+    event_handler.forget();
+
+    // Microphone is handled entirely in JavaScript
+
+    // Handle outgoing messages
+    let mut app_rx = app_rx.lock().await;
+    loop {
+        let message = app_rx.recv().await;
+        let Some(outgoing) = message else {
+            debug!("App pipe broken, exiting loop");
+            break;
+        };
+
+        let destinations = if let Some(address) = outgoing.recipient {
+            js_sys::Array::of1(&JsValue::from_str(&format!("{:#x}", address)))
+        } else {
+            js_sys::Array::new()
+        };
+
+        if let Err(e) = publish_data(
+            &room,
+            &outgoing.data,
+            !outgoing.unreliable,
+            destinations.into(),
+        )
+        .await
+        {
+            warn!("Failed to publish data: {:?}", e);
+            break;
+        }
+    }
+
+    close_room(&room)
+        .await
+        .map_err(|e| anyhow::anyhow!("Failed to close room: {:?}", e))?;
+
+    Ok(())
+}
+
+async fn handle_room_event(event: JsValue, transport_id: Entity, sender: Sender<PlayerUpdate>) {
+    // Parse the event from JavaScript
+    // This is a simplified version - you'll need to properly parse the JavaScript event object
+    if let Ok(event_type) = js_sys::Reflect::get(&event, &JsValue::from_str("type")) {
+        if let Some(event_type_str) = event_type.as_string() {
+            match event_type_str.as_str() {
+                "data_received" => {
+                    if let Ok(payload) = js_sys::Reflect::get(&event, &JsValue::from_str("payload"))
+                    {
+                        if let Ok(participant) =
+                            js_sys::Reflect::get(&event, &JsValue::from_str("participant"))
+                        {
+                            if let Ok(identity) =
+                                js_sys::Reflect::get(&participant, &JsValue::from_str("identity"))
+                            {
+                                if let Some(identity_str) = identity.as_string() {
+                                    if let Some(address) = identity_str.as_h160() {
+                                        // Convert payload to bytes
+                                        let data = js_sys::Uint8Array::new(&payload);
```

Comment:
not sure how serde will handle this, but it might work out of the box using Vec<u8> in the struct